src/ultk/language/semantics.py

```diff
@@ -50,20 +62,27 @@ def __hash__(self) -> int:
         return hash((self.name, tuple(self.__dict__)))
 
 
+@dataclass(frozen=True)
 class Universe:
 
     """The universe is the set of possible referent objects for a meaning."""
 
-    def __init__(self, referents: Iterable[Referent], prior: dict[str, float] = None):
-        self.referents = referents
-        self._referents_by_name = {referent.name: referent for referent in referents}
-        # set to uniform prior if none specified
-        size = len(referents)
-        prior = prior or {referent.name: 1 / size for referent in referents}
-        self.set_prior(prior)
+    referents: Iterable[Referent]
+    prior: dict[str, float] = None
+
+    @property
+    def _referents_by_name(self):
+        return {referent.name: referent for referent in self.referents}
 
-    def set_prior(self, prior: dict[str, float]):
-        self._prior = prior
+    @property
```

Comment:
I think `cached_property` from `functools` is even better here, so that it doesn't get recomputed every time (even though it's not super expensive for small sets of refs): https://docs.python.org/3/library/functools.html